osmomath/decimal.go

```diff
@@ -552,23 +560,49 @@ func (d BigDec) MustFloat64() float64 {
 // Dec returns the osmomath.Dec representation of a BigDec.
 // Values in any additional decimal places are truncated.
 func (d BigDec) Dec() Dec {
-	precisionDiff := PrecisionBigDec - PrecisionDec
-	precisionFactor := new(big.Int).Exp(big.NewInt(10), big.NewInt(int64(precisionDiff)), nil)
-
-	if precisionDiff < 0 {
-		panic("invalid decimal precision")
+	return d.DecWithPrecision(PrecisionDec)
+}
+
+// DecWithPrecision converts BigDec to Dec with desired precision
+// Example:
+// BigDec: 1.010100000000153000000000000000000000
+// precision: 4
+// Output Dec: 1.010100000000000000
+func (d BigDec) DecWithPrecision(precision int64) Dec {
+	var precisionFactor *big.Int
+	if precision > PrecisionDec {
+		panic(fmt.Sprintf("maximum Dec precision is (%v), provided (%v)", PrecisionDec, precision))
+	} else {
+		precisionFactor = precisionFactors[precision]
 	}
 
 	// Truncate any additional decimal values that exist due to BigDec's additional precision
 	// This relies on big.Int's Quo function doing floor division
 	intRepresentation := new(big.Int).Quo(d.BigInt(), precisionFactor)
 
 	// convert int representation back to SDK Dec precision
-	truncatedDec := NewDecFromBigIntWithPrec(intRepresentation, PrecisionDec)
+	truncatedDec := NewDecFromBigIntWithPrec(intRepresentation, precision)
 
 	return truncatedDec
 }
 
+// ChopPrecisionMut truncates all decimals after precision numbers after decimal point. Mutative
+// CONTRACT: precision <= PrecisionBigDec
+func (d *BigDec) ChopPrecisionMut(precision int64) BigDec {
+	precisionFactor := precisionFactors[precision]
+	// big.Quo truncates numbers that would have been after decimal point
+	d.i.Quo(d.i, precisionFactor)
+	d.i.Mul(d.i, precisionFactor)
+	return BigDec{d.i}
+}
+
+// ChopPrecision truncates all decimals after precision numbers after decimal point
+// CONTRACT: precision <= PrecisionBigDec
```

Comment:
What happens here if this CONTRACT has not been met? 👀

Reply:
Ah, it panics. Can we make this more explicit in godoc?

Reply:
I think we should add an explicit check and panic. Otherwise, it would panic with slice out of bounds which isn't very descriptive. Also, please cover by a test @pysel

Reply:
godocs updated:

[eadcaad](https://github.com/osmosis-labs/osmosis/pull/6261/commits/eadcaad29c36b21c4434cb2b6fc28519058069e4)

@pysel we still need to cover panic with a test